File: oo-workshop/tsexercises/src/domainmodeling/shoppingCartService.ts
import {CartItem} from './cartItem'
import {ShoppingCart} from './shoppingCart'
import {ShoppingCartRepository} from './shoppingCartRepository'

export class ShoppingCartService{

    private repository: ShoppingCartRepository

    public constructor(repository: ShoppingCartRepository){
        this.repository = repository
    }

    public addItemToShoppingCart(shoppingCartId: string, item: CartItem): ShoppingCart {
        const shoppingCart: ShoppingCart = this.repository.find(shoppingCartId)

        shoppingCart.items.push(item)

        this.repository.save(shoppingCart)

        return shoppingCart
    }

    public removeItemFromShoppingCart(shoppingCartId: string, itemId: string): ShoppingCart{
        const shoppingCart: ShoppingCart = this.repository.find(shoppingCartId)

        let deletionIndex: number = -1
        shoppingCart.items.forEach((item: CartItem, itemIndex) => {
            if(item.id === itemId){
                deletionIndex = itemIndex
            }
        })

        //the arbitrary looking '1' is saying how many items to remove starting at the deletion index
        shoppingCart.items.splice(deletionIndex, 1)
   
        return this.repository.save(shoppingCart)

    }

    public calculateShoppingCartTotal(shoppingCartId: string): number {
        const shoppingCart: ShoppingCart = this.repository.find(shoppingCartId)

        let total: number = 0
        shoppingCart.items.forEach((item: CartItem) => {
            total = total + item.price
        })
    
        return total
    }

    public getItemsInCart(shoppingCartId: string): CartItem[] {
        const shoppingCart: ShoppingCart = this.repository.find(shoppingCartId)

        return shoppingCart.items
    }
}

File: oo-workshop/mobprogrammingexercises/src/main/java/com/example/domainmodeling/anemic/ShoppingCartService.java
package com.example.domainmodeling.anemic;

import com.example.domainmodeling.anemic.solution.InvalidCustomerException;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class ShoppingCartService {

    private ShoppingCartRepository repository;

    public ShoppingCartService(ShoppingCartRepository repository) {
        this.repository = repository;
    }

    public ShoppingCart createShoppingCart(final ShoppingCart shoppingCart) throws InvalidCustomerException {
        ShoppingCartValidator.IsShoppingCartValid(shoppingCart);

        return this.repository.save(shoppingCart);
    }

    public ShoppingCart addItemToShoppingCart(final String shoppingCartId, final CartItem item) {
        final ShoppingCart shoppingCart = this.repository.find(shoppingCartId);

        shoppingCart.getItems().add((item));

        this.repository.save(shoppingCart);

        return shoppingCart;
    }

    public ShoppingCart removeItemFromShoppingCart(final String shoppingCartId, final String itemId) {
        final ShoppingCart shoppingCart = this.repository.find(shoppingCartId);

        final List<CartItem> cartWithRemovedItem = shoppingCart.getItems()
                .stream()
                .filter(cartItem -> {
                    return cartItem.getId() != itemId;
                }).collect(Collectors.toList());

        shoppingCart.setItems(cartWithRemovedItem);

        return this.repository.save(shoppingCart);

    }

    public BigDecimal calculateShoppingCartTotal(final String shoppingCartId) {
        final ShoppingCart shoppingCart = this.repository.find(shoppingCartId);

        BigDecimal total = BigDecimal.ZERO;
        for (final CartItem cartItem : shoppingCart.getItems()) {
            total = total.add(cartItem.getPrice());
        }

        return total;
    }

    public List<CartItem> getItemsInCart(final String shoppingCartId) {
        final ShoppingCart shoppingCart = this.repository.find(shoppingCartId);

        final List<CartItem> cartItems = shoppingCart.getItems()
                .stream()
                .sorted(Comparator.comparing(CartItem::getPrice))
                .collect(Collectors.toList());

        return cartItems;

    }
}
